src/Invio.Immutable/NullablePropertyHandler.cs

```diff
@@ -0,0 +1,119 @@
+            if (innerPropertyHandler == null) {
+                throw new ArgumentNullException(nameof(innerPropertyHandler));
+            }
+
+            this.innerPropertyHandler = innerPropertyHandler;
+        }
+
+        /// <remarks>
+        ///   This override handles nulls and then dispatches to the inner
+        ///   <see cref="IPropertyHandler"/> used to construct this instance.
+        /// </remarks>
+        /// <inheritdoc />
+        public override bool ArePropertyValuesEqual(object leftParent, object rightParent) {
+            if (leftParent == null) {
+                throw new ArgumentNullException(nameof(leftParent));
+            } else if (rightParent == null) {
+                throw new ArgumentNullException(nameof(rightParent));
+            }
+
+            var leftPropertyValue =
+                this.GetPropertyValueImplOrThrow(leftParent, nameof(leftParent));
+            var rightPropertyValue =
+                this.GetPropertyValueImplOrThrow(rightParent, nameof(rightParent));
+
+            // Handle nulls
+
+            if (Object.ReferenceEquals(leftPropertyValue, null)) {
+                return Object.ReferenceEquals(rightPropertyValue, null);
+            } else if (Object.ReferenceEquals(rightPropertyValue, null)) {
+                return false;
+            }
+
+            return this.innerPropertyHandler.ArePropertyValuesEqual(leftParent, rightParent);
+        }
+
+        /// <remarks>
+        ///   This override handles nulls and then dispatches to the inner
+        ///   <see cref="IPropertyHandler"/> used to construct this instance.
+        /// </remarks>
+        /// <inheritdoc />
+        public override int GetPropertyValueHashCode(object parent) {
+            if (parent == null) {
+                throw new ArgumentNullException(nameof(parent));
+            }
+
+            var propertyValue = this.GetPropertyValueImplOrThrow(parent, nameof(parent));
+
+            if (propertyValue == null) {
+                // Arbitrary (but prime!) number for null.
+
+                return 37;
+            }
+
+            return this.innerPropertyHandler.GetPropertyValueHashCode(parent);
+        }
+
+        /// <summary>
+        ///   This method is not supported.
+        /// </summary>
+        /// <inheritdoc />
+        /// <exception cref="NotSupportedException">
+        ///   This method is not supported.
+        /// </exception>
+        protected override bool ArePropertyValuesEqualImpl(TProperty? leftValue, TProperty? rightValue) {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        ///   This method is not supported.
+        /// </summary>
+        /// <inheritdoc />
+        /// <exception cref="NotSupportedException">
+        ///   This method is not supported.
+        /// </exception>
+        protected override int GetPropertyValueHashCodeImpl(TProperty? propertyValue) {
+            throw new NotSupportedException();
+        }
```

Comment:
I'd argue these are not needed. As part of inheritance, the class inheriting from the base implementation is responsible for all aggregated behavior across layers, versus composition where the contract between layers remains consistent with the interface. Tests are what will ensure it this code does not become relevant in some future context.

Reply:
My concern is that a code change to the base class introduces a new use case that invokes these methods. I have no way to implement these methods in any sane way, and if they got called with their default implementations it would result in some very confusingly unexpected behavior. Adding these exceptions makes it fail fast if these ever get called, which they really really shouldn't.